lib/flow/Particle.cpp

```diff
@@ -28,66 +27,55 @@ Particle::Particle( float x, float y, float z, float t, float val )
     value      = val;
 }
 
-void
-Particle::AttachProperty( float v )
+void Particle::AttachProperty( float v )
 {
-    auto itr = _properties.cbefore_begin();  
-    for( const auto& x : _properties )
-    {
-        (void) x;
-        ++itr;
-    }
-    _properties.insert_after( itr, v );
-    _nProperties++;
+    auto before_itr = _properties.cbefore_begin();  
+    for( auto itr = _properties.cbegin(); itr != _properties.cend(); ++itr )
+        ++before_itr;
+
+    _properties.insert_after( before_itr, v );
 }
 
-float
-Particle::RetrieveProperty( int idx ) const
+auto Particle::GetPropertyList() const -> const std::forward_list<float>&
 {
-    if( idx < 0 || idx > _nProperties )
-        throw std::out_of_range( "flow::Particle" );
-
-    auto itr = _properties.cbegin();
-    for( int i = 0; i < idx; i++ )
-    {
-        ++itr;
-    }
-    return *itr;
+    return _properties;
 }
 
-void
-Particle::ClearProperties()
+void Particle::ClearProperty()
 {
     _properties.clear();
-    _nProperties = 0;
 }
-
-int
-Particle::GetNumOfProperties() const
+    
+void Particle::RemoveProperty( size_t target_i )
 {
-    return (_nProperties);
+    size_t current_i = 0;
+    auto before_it = _properties.cbefore_begin();
+    for( auto it = _properties.cbegin(); it != _properties.cend(); ++it ) {
```

Comment:
This seems like awfully complicated code to perform what appears to be simply removing the ith member of a list. Is there a reason for this complexity, or am I missing something?

Reply:
I couldn't think of a simpler way to remove the ith member. Am I missing something?

Reply:
How about simply:

`_properties.erase_after(_properties.cbefore_begin()+target_i)`

Reply:
Hmm, it seems 
1. If `target_i` is out of range, it'll result in an undefined behavior? 
2. Are you sure this syntax works? It seems the iterator type of a forward_list is a `ForwardIterator`, which only supports self increment operators.  Operators such as `auto itr2 = itr1 + n` is something only a `RandomIterator` supports to my understanding. 

![image](https://user-images.githubusercontent.com/814885/98284820-ed169880-1f5e-11eb-8936-48a98fcbb628.png)

Reply:
If the offset it out of range I'd guess it behaves just like any other container. You can always check the offset against the size. 

Re whether it works: my understanding is that an offset added to an iterator results in another iterator: i.e. there is no difference between incrementing an iterator by 'n' steps or adding 'n' to an iterator: they both result in an iterator that is 'n' steps from wherever you started.

Reply:
Ok, no need to guess anymore. I tested that your suggested syntax will fail when compiling. 

The reason it fails is that it's a `ForwardIterator` rather than a `RandomAccessIterator`. The former is a subset of the latter. The behavior you described is a requirement of a `RandomAccessIterator` but not a requirement of a `ForwardIterator`.

Reply:
Makes sense. Now we know. Thanks for checking.